### PR TITLE
Issue #285 express-session free support

### PR DESCRIPTION
**Pull request by lovemaths**
We used to save state and nonce in session, so session is always required. Now we can encrypt and decrypt
state and nonce, and put them into cookie, so session is no longer mandatory.

We provided two more options: useCookieInsteadOfSession, cookieEncryptionKeys. To use cookie, first set the first option to true.
For the second option, provide an array of { key: '...', iv: '...' } items. For each item, key is any string
of size 32, and iv is any string of size 12. We will use the first item (with AES-256-GCM algorithm) to encrypt cookie, but we try every
item to decrypt cookie, so you can easily do key rollover.

### CODE REVIEW

File: test/End_to_end_test/oidc_cookie_test.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+var service = chromedriver.get_service();
+var webdriver = chromedriver.webdriver;
+var By = webdriver.By;
+var until = webdriver.until;
+var base64url = require('base64url');
+
+var chai = require('chai');
+var expect = chai.expect;
+
+const TEST_TIMEOUT = 500000; // 30 seconds
```
Comment (SomkaPe) [inline test/End_to_end_test/oidc_cookie_test.js]: looks like 500 seconds

**Comment by SomkaPe** [comment]
:shipit:

File: lib/cookieContentHandler.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+  if (!cookieEncryptionKeys || !Array.isArray(cookieEncryptionKeys) || cookieEncryptionKeys.length === 0)
+    throw new Error('CookieContentHandler: cookieEncryptionKeys must be a non-emptry array');
+
+  for (var i = 0; i < cookieEncryptionKeys.length; i++) {
+    var item = cookieEncryptionKeys[i];
+    if (!item.key || !item.iv)
+      throw new Error(`CookieContentHandler: array item ${i+1} in cookieEncryptionKeys must have the form { key: , iv: }`);
+    if (item.key.length !== 32)
+      throw new Error(`CookieContentHandler: key number ${i+1} is ${item.key.length} bytes, expected: 32 bytes`);
```
Comment (cadesalaberry) [inline lib/cookieContentHandler.js]: @lovemaths 

Why would the key need to be exactly 32 bytes ?

The [cookie-session](https://www.npmjs.com/package/cookie-session) seems to accept secret/keys of any size.

File: lib/oidcstrategy.js [hunk 3]
```diff
@@ -542,6 +573,7 @@ Strategy.prototype.authenticate = function authenticateStrategy(req, options) {
   var prompt = options && options.prompt;
   var extraAuthReqQueryParams = options && options.extraAuthReqQueryParams;
   var extraTokenReqQueryParams = options && options.extraTokenReqQueryParams;
+  var response = options && options.response;
```
Comment (cadesalaberry) [inline lib/oidcstrategy.js]: @lovemaths `options.response` is always undefined in my case.

In express 4, response can be accessed with `req.res`. Was this PR developed with this in mind ?

**Comment by frantzyy** [comment]
Would it be possible to see an example of how to use this feature? I've searched everywhere and haven't found one. I understand how to setup up the config with the exception of do I need to specifically call the /token endpoint  or will passport-azure-ad do this for me. If so where will the jwt/token be available? in the strategy callback or response of the redirect?

**Pull request merged**